src/ucp/core/ucp_ep.c

```diff
@@ -86,9 +86,12 @@ ucs_status_t ucp_ep_new(ucp_worker_h worker, uint64_t dest_uuid,
             goto err_free_ep;
         }
 
+        /* NOTE: both queues are in the union must be the same type in order to
+         *       initialize once */
+        UCS_STATIC_TYPE_MATCH(ep->ext.stream->data, ep->ext.stream->reqs);
```

Comment:
why need to initialize once? just make sure am_data_len selects the one which is initialized..

Reply:
Let's say, the queues have different types (i.e. queue and list) and stored in union. Then it's need to re-initialize one from them every time when first entry should be inserted. Since they are the same type, empty container is valid for both.

Reply:
i see.. but it's a bit confusing/unclean.. how about having just one queue field, and let the "am_data_len" specify which kind of *objects* are present on that queue? since there is also assumption here that after one queue becomes empty, the other one would be already initialized

Reply:
I thought about it too, but it also may confuse when the same queue contains different objects (in different time). if you see this way more clear - choose the name :) pls note we are typically naming queue/list fields for storing in the queue/list but not the container instance name.

Reply:
i think single queue is more clear because the variables are not re-initialized after switching the role. could call it `match_q` or just `queue`